lib/puppet/functions/new.rb

```diff
@@ -14,10 +14,7 @@
 
   def new_instance(scope, t, *args)
     return args[0] if args.size == 1 && !t.is_a?(Puppet::Pops::Types::PInitType) && t.instance?(args[0])
-    result = catch :undefined_value do
-      new_function_for_type(t, scope).call(scope, *args)
-    end
-    assert_type(t, result)
+    result = assert_type(t, new_function_for_type(t, scope).call(scope, *args))
```

Comment:
the catching of :undefined_value is removed - there is still logic that throws that and I think the result becomes  a bad error (unhandled throw of :undefied_value) instead of the error produced by assert_type.

Reply:
Where is the logic that still throws that? I thought I had removed all of it.